src/core/lib/event_engine/ares_resolver.cc

```diff
@@ -0,0 +1,714 @@
+      ares_backup_poll_alarm_handle_.reset();
+    }
+    for (const auto& fd_node : fd_node_list_) {
+      if (!fd_node->already_shutdown) {
+        GRPC_ARES_RESOLVER_TRACE_LOG("request: %p shutdown fd: %s", this,
+                                     fd_node->polled_fd->GetName());
+        fd_node->polled_fd->ShutdownLocked(
+            absl::CancelledError("AresResolver::Orphan"));
+        fd_node->already_shutdown = true;
+      }
+    }
+  }
+  Unref(DEBUG_LOCATION, "Orphan");
+}
+
+void AresResolver::LookupHostname(
+    absl::string_view name, absl::string_view default_port,
+    EventEngine::DNSResolver::LookupHostnameCallback callback) {
+  absl::string_view host;
+  absl::string_view port_s;
+  if (!grpc_core::SplitHostPort(name, &host, &port_s)) {
+    event_engine_->Run(
+        [callback = std::move(callback),
+         status = absl::InvalidArgumentError(absl::StrCat(
+             "Unparseable name: ", name))]() mutable { callback(status); });
+    return;
+  }
+  GPR_ASSERT(!host.empty());
+  if (port_s.empty()) {
+    if (default_port.empty()) {
+      event_engine_->Run([callback = std::move(callback),
+                          status = absl::InvalidArgumentError(absl::StrFormat(
+                              "No port in name %s or default_port argument",
+                              name.data()))]() mutable { callback(status); });
+      return;
+    }
+    port_s = default_port;
+  }
+  int port = 0;
+  if (port_s == "http") {
+    port = 80;
+  } else if (port_s == "https") {
+    port = 443;
+  } else if (!absl::SimpleAtoi(port_s, &port)) {
+    event_engine_->Run([callback = std::move(callback),
+                        status = absl::InvalidArgumentError(absl::StrCat(
+                            "Failed to parse port in name: ",
+                            name))]() mutable { callback(status); });
+    return;
+  }
+  // TODO(yijiem): Change this when refactoring code in
+  // src/core/lib/address_utils to use EventEngine::ResolvedAddress.
+  grpc_resolved_address addr;
+  const std::string hostport = grpc_core::JoinHostPort(host, port);
+  if (grpc_parse_ipv4_hostport(hostport.c_str(), &addr,
+                               false /* log errors */) ||
+      grpc_parse_ipv6_hostport(hostport.c_str(), &addr,
+                               false /* log errors */)) {
+    // Early out if the target is an ipv4 or ipv6 literal.
+    std::vector<EventEngine::ResolvedAddress> result;
+    result.emplace_back(reinterpret_cast<sockaddr*>(addr.addr), addr.len);
+    event_engine_->Run(
+        [callback = std::move(callback), result = std::move(result)]() mutable {
+          callback(std::move(result));
+        });
+    return;
+  }
+  grpc_core::MutexLock lock(&mutex_);
+  callback_map_.emplace(++id_, std::move(callback));
+  auto* resolver_arg = new HostnameQueryArg(this, id_, name, port);
+  if (IsIpv6LoopbackAvailable()) {
+    ares_gethostbyname(channel_, std::string(host).c_str(), AF_UNSPEC,
```

Comment:
Now that we're using AF_UNSPEC, can we remove the address_sorting library?

Reply:
I hope so, maybe after we roll out the EE version and verify it works and remove the original `grpc_ares_wrapper` code. FWIW, the c-ares' addressing sorting looks very similar to the address_sorting library (they both implement RFC 6724), but seems to have some slight differences in rule ordering. Not sure if that matters though.

c-ares: https://source.corp.google.com/piper///depot/google3/third_party/ares/src/lib/ares__sortaddrinfo.c;l=282-286;rcl=540440028

address_sorting: https://source.corp.google.com/piper///depot/google3/third_party/address_sorting/address_sorting.c;l=300;bpv=0;bpt=1;rcl=543911652

Reply:
My apologies, I checked again and the order appears to be the same. The comment in `third_party/addressing_sorting` is a bit confusing and made me think that they were different.